acl.go

```diff
@@ -0,0 +1,110 @@
+package imap
+
+import (
+	"fmt"
+	"strings"
+)
+
+// IMAP4 ACL extension (RFC 2086)
+
+type RightSet string
+
+type Right byte
+
+const (
+	RightLookup     = Right('l') // mailbox is visible to LIST/LSUB commands
+	RightRead       = Right('r') // SELECT the mailbox, perform CHECK, FETCH, PARTIAL, SEARCH, COPY from mailbox
+	RightSeen       = Right('s') // keep seen/unseen information across sessions (STORE SEEN flag)
+	RightWrite      = Right('w') // STORE flags other than SEEN and DELETED
+	RightInsert     = Right('i') // perform APPEND, COPY into mailbox
+	RightPost       = Right('p') // send mail to submission address for mailbox, not enforced by IMAP4 itself
+	RightCreate     = Right('c') // CREATE new sub-mailboxes in any implementation-defined hierarchy
+	RightDelete     = Right('d') // STORE DELETED flag, perform EXPUNGE
+	RightAdminister = Right('a') // perform SETACL
+
+	AllRights = RightSet("lrswipcda")
+)
+
+type RightsIdentifier string
+
+const RightsIdentifierAnyone = RightsIdentifier("anyone")
+
+type RightModification byte
+
+const (
+	RightModificationReplace = RightModification(0)
+	RightModificationAdd     = RightModification('+')
+	RightModificationRemove  = RightModification('-')
+)
+
+// NewRights converts rights string into RightModification and RightSet with validation
+func NewRights(rights string, ignoreUnsupported bool) (RightModification, RightSet, error) {
+	rm := RightModificationReplace
+
+	if len(rights) == 0 {
+		return rm, "", nil
+	}
+
+	if rights[0] == byte(RightModificationAdd) || rights[0] == byte(RightModificationRemove) {
+		rm = RightModification(rights[0])
+		rights = rights[1:]
+	}
+
+	var newRights RightSet
+	for _, r := range rights {
+		if IsSupportedRight(r) {
+			newRights += RightSet(r)
+		} else if !ignoreUnsupported {
+			return rm, "", fmt.Errorf("unsupported right: '%v'", r)
+		}
+	}
+
+	return rm, RightSet(newRights), nil
+}
+
+func IsSupportedRight(r rune) bool {
```

Comment:
Maybe `IsStandardRight` would be a better name?

What is the use-case for this? Is it for server implementations? If so, maybe `IsStandardRightSet` taking a `RightSet` would be easier to use?

But I wonder if it's useful to expose this, server implementations will need to go through each right they receive anyways so it should be trivial for them to reject rights they don't support…

Reply:
I wanted to use this to filter nonstandard rights in NewRights function. You are right, this shouldn't be exposed.